### PR TITLE
[BEAM-2732] Metrics rely on statesampler state

### PR DESCRIPTION
**Pull request by pabloem**
This PR depends on PR #4375.

This PR is a backward-compatible change that allows metrics to rely on statesampler state to provide the current container.

This change should enable all clients to stop tracking metrics with its own context, and start using statesampler for that.

Benchmarks showed no impact on performance. Benchmark done using an ML Criteo pipeline that processes 10gb.

Results Before Change (5 runs):
Mean: 1809.2s, Stdev: 227.1622768

Results With Change (5 runs):
Mean: 1758.4s, Stdev: 93.46282684

### CODE REVIEW

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
r: @robertwb 
Here's the goal for the statesampler changes. To have everything rely on the same functionality to track context. This change is backwards compatible, but I'll make sure to port all clients immediately, so the code won't remain in this state for more than a couple weeks.

**Comment by pabloem** [comment]
* There's also the issue of step names for the operation. Right now I'm simply passing step_name in, which seems safe, but I'm not sure.

File: sdks/python/apache_beam/transforms/util.py [hunk 1]
```diff
@@ -221,6 +220,10 @@ def __init__(self,
     self._clock = clock
     self._data = []
     self._ignore_next_timing = False
+
+    # TODO(pabloem) - This should rely on internal counters. Not user metrics.
+    from apache_beam.metrics import Metrics
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: Why this change?

File: sdks/python/apache_beam/transforms/util.py [hunk 2]
```diff
@@ -221,6 +220,10 @@ def __init__(self,
     self._clock = clock
     self._data = []
     self._ignore_next_timing = False
+
+    # TODO(pabloem) - This should rely on internal counters. Not user metrics.
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: No, this is a user-level transform.

File: sdks/python/apache_beam/metrics/execution.py [hunk 3]
```diff
@@ -135,18 +136,40 @@ def set_metrics_supported(self, supported):
...
     self.set_container_stack()
     index = len(self.PER_THREAD.container) - 1
     if index < 0:
       return None
     return self.PER_THREAD.container[index]
 
+  def current_container(self):
+    """Returns the current MetricsContainer."""
+    sampler = statesampler.EXECUTION_STATE_SAMPLERS.current_sampler()
+    if sampler is None:
```
Comment (robertwb) [inline sdks/python/apache_beam/metrics/execution.py]: When would this be false? Can we simply assert it'll be present after your CL rather than keeping both the old and new paths in parallel?

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 4]
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
     self.sampling_period_ms = sampling_period_ms
     super(StateSampler, self).__init__(sampling_period_ms)
 
+    # TODO(pabloem) - Remove this once all clients register the
+    # sampler independently.
+    self.register()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: As I mentioned before, I'd really like to avoid being in this state, even temporarily.

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 5]
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
...
+  def simple_tracker():
+    sampler = StateSampler('', CounterFactory())
+    sampler.register()
+    return sampler
+
+  @staticmethod
+  def create_and_register_tracker(
+      prefix, counter_factory, sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
+    sampler = StateSampler(prefix, counter_factory, sampling_period_ms)
+    sampler.register()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Won't this be a double registration?

File: sdks/python/apache_beam/runners/worker/statesampler_slow.py [hunk 6]
```diff
@@ -49,19 +55,15 @@ def start(self):
...
 
 class ScopedState(object):
 
-  def __init__(self, sampler, name, counter=None):
+  def __init__(self, sampler, name, counter=None, metrics_container=None):
     self.state_sampler = sampler
     self.name = name
     self.counter = counter
     self.nsecs = 0
+    self.metrics_container = metrics_container
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler_slow.py]: Adding a counter container here seems out-of-place, and who knows what else might get added.

Instead, I'd suggest an architecture where the there is the ability to get the local state (which is an object that can be used as a key and has basic information like step name) which other frameworks (e.g. metrics) consume in a self-contained way. This also reduces the scope of the global state we must store to access these contexts implicitly from user code.

Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/statesampler_slow.py]: I agree that tagging state here may be useful for cleanup when the state is no longer needed, so I suppose this is OK.

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 7]
```diff
@@ -29,6 +31,22 @@
...
+  """ Per-thread state sampler. """
+  def __init__(self):
+    super(ExecutionStateSamplers, self).__init__()
+    self._current_sampler = None
+
+  def current_sampler(self):
+    return self._current_sampler
+
+  def set_sampler(self, sampler):
+    self._current_sampler = sampler
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Throw an error if the sampler was already set.

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 8]
```diff
@@ -29,6 +31,22 @@
...
+
+
+EXECUTION_STATE_SAMPLERS = ExecutionStateSamplers()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Perhaps the name should indicate that this is per-thread? In general, creating a global singleton class rather than module-level functions is a Javaism.

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 9]
```diff
@@ -29,6 +31,22 @@
...
 
 
+class ExecutionStateSamplers(threading.local):
+  """ Per-thread state sampler. """
+  def __init__(self):
+    super(ExecutionStateSamplers, self).__init__()
+    self._current_sampler = None
+
+  def current_sampler(self):
+    return self._current_sampler
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Throw an error if it was not set?

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 10]
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
...
+  @staticmethod
+  def create_and_register_tracker(
+      prefix, counter_factory, sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
+    sampler = StateSampler(prefix, counter_factory, sampling_period_ms)
+    sampler.register()
+    return sampler
+
+  def register(self):
+    EXECUTION_STATE_SAMPLERS.set_sampler(self)
+    self._registered = True
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Is this value very queried?

File: sdks/python/apache_beam/runners/worker/statesampler.py [hunk 11]
```diff
@@ -49,18 +67,43 @@ def __init__(self, prefix, counter_factory,
...
+    return sampler
+
+  @staticmethod
+  def create_and_register_tracker(
+      prefix, counter_factory, sampling_period_ms=DEFAULT_SAMPLING_PERIOD_MS):
+    sampler = StateSampler(prefix, counter_factory, sampling_period_ms)
+    sampler.register()
+    return sampler
+
+  def register(self):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/statesampler.py]: Maybe register_on_current_thread? I'd actually have the caller call EXECUTION_STATE_SAMPLERS.set_sampler directly, so the interaction of getting the sampler is symmetrc with setting it.

**Comment by pabloem** [comment]
Let me know if PR 4504 seems like a more sound implementation, or if PR 4387 (this one) seems more sensible.

**Comment by pabloem** [comment]
Changes from the current commit:
1. Addet get/set_current_tracker methods
2. Removed unnecessary ExecutionStateTrackers class.
3. Removed register() call from StateSampler constructor. Instead, a state sampler registers itself when start() is called (similar to java), or a user can use set_current_tracker
4. Removed simple_tracker, and make_tracker calls. StateSampler should be created via its constructor.

Regarding your questions:
1. The difference between a counter and a metric cell? 
- A metric cell manages locking and commit state of metrics.
- Counters created by a CounterFactory (I believe you were referring to those?) are used internally only.

2. Is the thread-local variable accessed for every metric update?
Yes, that is the current state (currently, and it will not change in this PR). Definitely an optimization we could consider.

**Comment by pabloem** [comment]
Rebased

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
@robertwb PTAL

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
All integration tests but `test_streaming_wordcount_it` are passing. I'll try to fix that.....

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
Python Postcommit tests are passing. I'm not sure why the `mvn clean install -pl sdks/python -am -am...` are broken, as they are a subset of the Postcommit suite:

image

@robertwb PTAL

**Comment by pabloem** [comment]
Rebased change.

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
@robertwb PTAL

**Comment by pabloem** [comment]
Rebased change.

**Comment by pabloem** [comment]
Run Python PostCommit.

**Comment by pabloem** [comment]
Retest this please

**Comment by pabloem** [comment]
Retest this please.

**Comment by pabloem** [comment]
Run Python PostCommit

File: sdks/python/apache_beam/runners/direct/direct_runner.py [hunk 12]
```diff
@@ -338,6 +337,7 @@ def run_pipeline(self, pipeline):
     from apache_beam.runners.direct.transform_evaluator import \
       TransformEvaluatorRegistry
     from apache_beam.testing.test_stream import TestStream
+    from apache_beam.metrics.execution import MetricsEnvironment
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/direct_runner.py]: Undo this change? Metrics should not be changed to depend on anything in the runners package. (Either that or the lazy import should be made there.)

File: sdks/python/apache_beam/runners/direct/executor.py [hunk 13]
```diff
@@ -290,70 +293,87 @@ def __init__(self, transform_evaluator_registry, evaluation_context,
...
+                   finish_state):
+    """Attempts to run a bundle. Called within the 'start' ExecutionState."""
     evaluator = self._transform_evaluator_registry.get_evaluator(
         self._applied_ptransform, self._input_bundle,
-        side_input_values, scoped_metrics_container)
+        side_input_values)
 
-    with scoped_metrics_container:
-      evaluator.start_bundle()
+    evaluator.start_bundle()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/direct/executor.py]: Why is this not called from within start_state?

File: sdks/python/apache_beam/runners/worker/operations.py [hunk 14]
```diff
@@ -133,24 +133,25 @@ def __init__(self, name_context, spec, counter_factory, state_sampler):
 
     # These are overwritten in the legacy harness.
     self.metrics_container = MetricsContainer(self.name_context.metrics_name())
-    self.scoped_metrics_container = ScopedMetricsContainer(
-        self.metrics_container)
+    self.scoped_metrics_container = ScopedMetricsContainer()
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/worker/operations.py]: Can this just be deleted?

File: sdks/python/apache_beam/transforms/util.py [hunk 15]
```diff
@@ -221,6 +220,9 @@ def __init__(self,
     self._clock = clock
     self._data = []
     self._ignore_next_timing = False
+
+    from apache_beam.metrics import Metrics
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/util.py]: Undo this change. Transforms should be free to use metrics (at least the public metrics write API).

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 12 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: Or perhaps don't import metrics.execution when importing the public (write) Metrics api.

File: sdks/python/apache_beam/runners/direct/executor.py (hunk 13 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/direct/executor.py]: The function `attempt_call` is already called within start_state (because side input initialization happens before `attempt_call` is called. I've documented this in the `attempt_call` docstring, but we could do it differently if you think that'd make more sense.

File: sdks/python/apache_beam/runners/worker/operations.py (hunk 14 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/worker/operations.py]: Right. Done.

File: sdks/python/apache_beam/transforms/util.py (hunk 15 above)
Comment (pabloem) [reply sdks/python/apache_beam/transforms/util.py]: Done.

File: sdks/python/apache_beam/runners/direct/direct_runner.py (hunk 12 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/direct/direct_runner.py]: That is reasonable, and I agree. I've fixed this by having a lazy import in the metrics execution module.
Currently, we're working under the idea that state sampler is the global context provider, so metrics rely on it. I'd think that makes it into a module that provides a service to other modules that require/handle context management. Though I'd agree that a more loose coupling would be good to have.

**Comment by pabloem** [comment]
Thanks Robert!

File: sdks/python/apache_beam/runners/worker/operations.py (hunk 14 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/worker/operations.py]: Ah nevermind. The op.scoped_metrics_container is used by Dataflow worker code. Needs to be available.

**Comment by pabloem** [comment]
Run Python PostCommit

File: sdks/python/apache_beam/runners/worker/operations.py (hunk 14 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/worker/operations.py]: OK. Please add a comment (with a JIRA) here about that.

File: sdks/python/apache_beam/runners/direct/executor.py (hunk 13 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/direct/executor.py]: Yes, I think it'd be cleaner to have all states explicitly entered within attempt_call than make this an implicit assumption about the caller.

Comment (pabloem) [reply sdks/python/apache_beam/runners/direct/executor.py]: Done.

File: sdks/python/apache_beam/runners/worker/operations.py (hunk 14 above)
Comment (pabloem) [reply sdks/python/apache_beam/runners/worker/operations.py]: Done. Thanks Robert.

**Comment by pabloem** [comment]
Retest this please.

**Comment by pabloem** [comment]
Squashed commits. Letting tests run.

**Comment by robertwb** [comment]
Jenkins: retest this please

**Comment by pabloem** [comment]
PreCommits seem to be broken, also on my machine after I rebased.

**Comment by robertwb** [comment]
What errors were you getting on your machine?

**Comment by pabloem** [comment]
Monsterpaste:

```
======================================================================
ERROR: test_basics_with_type_check (apache_beam.examples.cookbook.group_with_coder_test.GroupWithCoderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder_test.py", line 53, in test_basics_with_type_check
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder.py", line 118, in run
    | WriteToText(known_args.output))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpJcX3Kr*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics_without_type_check (apache_beam.examples.cookbook.group_with_coder_test.GroupWithCoderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder_test.py", line 74, in test_basics_without_type_check
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/group_with_coder.py", line 118, in run
    | WriteToText(known_args.output))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpisQlPp*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_multiple_output_pardo (apache_beam.examples.cookbook.multiple_output_pardo_test.MultipleOutputParDo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/multiple_output_pardo_test.py", line 56, in test_multiple_output_pardo
    '--output=%s' % result_prefix])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/cookbook/multiple_output_pardo.py", line 154, in run
    lines = p | ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpnmwepq*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_debugging_test.WordCountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_debugging_test.py", line 52, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_debugging.py", line 134, in run
    | 'FilterText' >> beam.ParDo(FilterTextFn('Flourish|stomach')))
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpSmPq3X*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_minimal_test.WordCountMinimalTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_minimal_test.py", line 47, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_minimal.py", line 100, in run
    lines = p | ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmprNRvNX*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}

======================================================================
ERROR: test_basics (apache_beam.examples.wordcount_test.WordCountTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount_test.py", line 46, in test_basics
    '--output=%s.result' % temp_path])
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/examples/wordcount.py", line 90, in run
    lines = p | 'read' >> ReadFromText(known_args.input)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 522, in __init__
    skip_header_lines=skip_header_lines)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/textio.py", line 117, in __init__
    validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpnAxeZ2*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}
-------------------- >> begin captured logging << --------------------
root: INFO: Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.
--------------------- >> end captured logging << ---------------------

======================================================================
ERROR: test_sink_transform (apache_beam.io.avroio_test.TestAvro)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/avroio_test.py", line 390, in test_sink_transform
    readback = p | avroio.ReadFromAvro(path + '*') | beam.Map(json.dumps)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/avroio.py", line 137, in __init__
    self._source = _AvroSource(file_pattern, min_bundle_size, validate=validate)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 119, in __init__
    self._validate()
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/options/value_provider.py", line 124, in _f
    return fnc(self, *args, **kwargs)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filebasedsource.py", line 176, in _validate
    match_result = FileSystems.match([pattern], limits=[1])[0]
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystems.py", line 166, in match
    return filesystem.match(patterns, limits)
  File "/usr/local/google/home/pabloem/codes/global-sampler-metrics/sdks/python/apache_beam/io/filesystem.py", line 600, in match
    raise BeamIOError("Match operation failed", exceptions)
BeamIOError: Match operation failed with exceptions {'/tmp/tmpN4Vb7G*': BeamIOError("List operation failed with exceptions {'/tmp': OSError(2, 'No such file or directory')}",)}

```

**Comment by aaltay** [comment]
https://github.com/apache/beam/pull/5167 should address those errors.

**Comment by pabloem** [comment]
Run Python PostCommit

**Comment by pabloem** [comment]
I believe this is ready to be merged.
